### PR TITLE
Fix Rouchon1_5 and implement caching

### LINKED ISSUE #78: Fix `inv_sqrtm` (used by `MERouchon1_5`)
https://github.com/pierreguilmin/torchqdynamics/blob/5d797a4ca199af16f0689fade49046a4452fba07/torchqdynamics/mesolve/rouchon.py#L107-L108

### PR DESCRIPTION
**Pull request by gautierronan**
Fixes #78.

### CODE REVIEW

File: dynamiqs/mesolve/rouchon.py [hunk 1]
```diff
@@ -91,13 +91,27 @@ def backward_augmented(self, t: float, rho: Tensor, phi: Tensor):
...
 
-        Note:
-            No need for trace renormalization since the scheme is trace-preserving
-            by construction.
+        # compute the inverse square root renormalization matrix
+        # (for time-dependent Hamiltonians, exclude the Hamiltonian from normalization)
+        if isinstance(self.H, ConstantTDTensor):
+            H_nh = self.H_nh(0.0)  # (b_H, 1, n, n)
+        elif isinstance(self.H, CallableTDTensor):
+            H_nh = -0.5j * self.sum_no_jump  # (b_H, 1, n, n)
```
Comment (pierreguilmin) [inline dynamiqs/mesolve/rouchon.py:102]: Where did you see this? Are you sure it's ok to exclude the Hamiltonian from the normalisation?

**Review by pierreguilmin — CHANGES_REQUESTED** [review]
Verrryyy nice thanks a lot!!

File: dynamiqs/mesolve/rouchon.py (hunk 1 above)
Comment (gautierronan) [reply dynamiqs/mesolve/rouchon.py:102]: Yes essentially it's better than nothing. Recomputing the normalization at every time step would be extremely costly, and it's often the case that numerical divergence is caused by the Ldag @ L terms that explode as the Fock truncature gets larger.

**Comment by gautierronan** [comment]
NB: We might want to merge `Rouchon1` and `Rouchon1_5` at some point because they share the exact same `forward` and `backward_augmented` implementations. The only difference is the init. Plus, they're actually both order 1 methods, so we could just add an option to not add the square root normalization, or even completely remove the current `Rouchon1` method.

File: dynamiqs/solvers/utils/utils.py [hunk 2]
```diff
@@ -45,7 +45,8 @@ def inv_sqrtm(mat: Tensor) -> Tensor:
          https://github.com/pytorch/pytorch/issues/25481#issuecomment-584896176.
     """
```
Comment (pierreguilmin) [inline dynamiqs/solvers/utils/utils.py:46]:
```diff
- """Compute the inverse square root of a matrix using its eigendecomposition.
+ """Compute the inverse square root of a complex Hermitian or real symmetric matrix
+    using its eigendecomposition.
```

File: dynamiqs/mesolve/rouchon.py [hunk 3]
```diff
@@ -91,13 +88,35 @@ def backward_augmented(self, t: float, rho: Tensor, phi: Tensor):
...
-        Note:
-            No need for trace renormalization since the scheme is trace-preserving
-            by construction.
+        # compute the inverse square root renormalization matrix
+        # (for time-dependent Hamiltonians, exclude the Hamiltonian from normalization)
+        # H_nh, M0, M0rev, self.S, self.Srev: (b_H, 1, n, n)
+        if isinstance(self.H, ConstantTDTensor):
+            H_nh = self.H_nh(0.0)
+        elif isinstance(self.H, CallableTDTensor):
+            H_nh = -0.5j * self.sum_no_jump
```
Comment (pierreguilmin) [inline dynamiqs/mesolve/rouchon.py:100]: Can we call this `H_nh_tmp` as it is not the real `H_nh`?

**Review by pierreguilmin — APPROVED** [review]
Minor comments.

> NB: We might want to merge Rouchon1 and Rouchon1_5 at some point because they share the exact same forward and backward_augmented implementations. The only difference is the init. Plus, they're actually both order 1 methods, so we could just add an option to not add the square root normalization, or even completely remove the current Rouchon1 method.

I'm down to merge them (here or in another PR), but we want to leep keep the current Rouchon1 method for two reasons:
- to compare when benchmarking (and understand how imptt is the normalisation)
- for sparse matrices, the renormalization destroys the sparness

**Review by pierreguilmin — APPROVED** [review]

**Pull request merged**